### PR TITLE
User expiration

### PR DESCRIPTION
**Pull request by cristianames92**
## Summary

This PR adds expiration to the existing token

## Known Issues

none

## Trello Card

https://trello.com/c/KTlx2qtv/38-hacer-que-una-sesion-expire

### CODE REVIEW

File: app/tools/jwtToken.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+const jwt = require('jsonwebtoken'),
+  JWT_KEY = require('../constants').jwt_key;
+
+exports.createToken = (payload, exp = Date.now() + Math.floor(process.env.SESSION_EXP) * 1000) => {
```
Comment (gonziesc) [inline app/tools/jwtToken.js]: use momentjs, it's the best package for managing dates stuff, i know this case is simple but i prefer that you start using it

File: app/tools/jwtToken.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      } else {
+        reject(jwtError);
+      }
+    });
+  });
+};
+
+exports.veryfyExpiration = (payload, comparisonTimestamp = Date.now()) => {
+  return new Promise((resolve, reject) => {
+    if (payload.exp < Date.now()) {
```
Comment (gonziesc) [inline app/tools/jwtToken.js]: same here

File: test/controllers/users.js [hunk 3]
```diff
@@ -204,7 +204,7 @@ describe('users controller', () => {
       );
     });
     context('When requesting with a valid token', () => {
-      const validToken = jwt.sign({}, JWT_KEY);
+      const validToken = jwt.createToken({}, Date.now() + 60000);
```
Comment (gonziesc) [inline test/controllers/users.js]: same here

**Review by gonziesc — CHANGES_REQUESTED** [review]

File: app/tools/jwtToken.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+const jwt = require('jsonwebtoken'),
```
Comment (gonziesc) [inline app/tools/jwtToken.js:1]: we dont use tools, we use helpers, but its ok for now

File: app/tools/jwtToken.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      if (decoded) {
+        resolve(decoded);
+      } else {
+        reject(jwtError);
+      }
+    });
+  });
+};
+
+exports.veryfyExpiration = (payload, comparisonTimestamp = moment()) => {
```
Comment (gonziesc) [inline app/tools/jwtToken.js:24]: i wont reject a promise for this. just return the value and eval with a if, or return a boolean but resolving. Promises should fail and enter in a catch when the external dependency fails

**Review by gonziesc — CHANGES_REQUESTED** [review]

File: app/tools/jwtToken.js (hunk 4 above)
Comment (cristianames92) [reply app/tools/jwtToken.js:1]: 👍

File: app/tools/jwtToken.js (hunk 5 above)
Comment (cristianames92) [reply app/tools/jwtToken.js:24]: resolving an object with value and error should be fine?

Comment (gonziesc) [reply app/tools/jwtToken.js:24]: yes!

File: app/controllers/users.js [hunk 6]
```diff
@@ -46,7 +46,20 @@ exports.list = (req, res, next) => {
   return jwt
     .verifyToken(req.headers.session_token)
     .then(jwt.veryfyExpiration)
-    .then(() => {
+    .then(payload => {
+      if (!payload.value) {
+        if (payload.error) {
+          console.log(`------------- H - ${payload.error}`);
```
Comment (gonziesc) [inline app/controllers/users.js]: delete console logs

File: app/tools/jwtToken.js [hunk 7]
```diff
@@ -12,20 +12,29 @@ exports.verifyToken = token => {
   return new Promise((resolve, reject) => {
     jwt.verify(token, JWT_KEY, (jwtError, decoded) => {
       if (decoded) {
-        resolve(decoded);
+        resolve({ value: decoded });
       } else {
-        reject(jwtError);
+        resolve({ error: jwtError });
```
Comment (gonziesc) [inline app/tools/jwtToken.js]: revert this change, here it was ok to reject

File: app/tools/jwtToken.js [hunk 8]
```diff
@@ -12,20 +12,29 @@ exports.verifyToken = token => {
...
+      } else {
+        reject('Unknown error');
+      }
+      return; 
+    }
+    if (comparisonTimestamp.isAfter(moment(payload.value.exp || 0))) {
+      resolve({ error: `Expired token - exp: ${payload.value.exp}` });
     } else {
-      resolve(payload.data);
+      resolve({ value: payload.value.data });
```
Comment (gonziesc) [inline app/tools/jwtToken.js]: i would always return the same object {value, error}, but its ok

**Review by gonziesc — CHANGES_REQUESTED** [review]

File: app/tools/jwtToken.js (hunk 8 above)
Comment (cristianames92) [reply app/tools/jwtToken.js]: 👍

**Review by gonziesc — APPROVED** [review]
